polaris-service/src/main/java/org/apache/polaris/service/admin/PolarisServiceImpl.java

```diff
@@ -505,20 +515,23 @@ public Response addGrantToCatalogRole(
               catalogName, catalogRoleName, Namespace.of(namespaceParts), privilege);
           break;
         }
-      case CatalogGrant catalogGrant:
+        case CatalogGrant catalogGrant:
         {
           PolarisPrivilege privilege =
               PolarisPrivilege.valueOf(catalogGrant.getPrivilege().toString());
           adminService.grantPrivilegeOnCatalogToRole(catalogName, catalogRoleName, privilege);
           break;
         }
-      default:
-        LOGGER
-            .atWarn()
-            .addKeyValue("catalog", catalogName)
-            .addKeyValue("role", catalogRoleName)
-            .log("Don't know how to handle privilege grant: {}", grantRequest);
-        return Response.status(Response.Status.BAD_REQUEST).build();
+        default:
+          LOGGER
+              .atWarn()
+              .addKeyValue("catalog", catalogName)
+              .addKeyValue("role", catalogRoleName)
+              .log("Don't know how to handle privilege grant: {}", grantRequest);
+          return Response.status(Response.Status.BAD_REQUEST).build();
+      }
+    } catch (IllegalStateException e) {
+      throw new AlreadyExistsException("Grant already exists or resolution failed");
```

Comment:
If we throw `jakarta.persistence.EntityExistsException` instead we won't need to assume that `IlegalStateException` should translate to `AlreadyExistsException`. In particular, some elements of the MetaStoreManger very well may throw `IllegalStateException` for true illegal states that are server errors and wouldn't make sense to translate into `AlreadyExistsException`.